File: C++/LongestValidParentheses.cpp
#include<iostream>
#include<vector>
#include<stack>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
class Node
{
public:
	char ch;
	int index;
	Node(char c, int i)
	{
		ch = c;
		index = i;
	}

};
class Solution {
public:
    int longestValidParentheses(string s) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
		int len = s.size();
		if(!len)
			return 0;
		stack<Node> nodes;
		nodes.push(Node(')',-1));
		int ret = 0;
		for(int i = 0; i < len; i++)
		{
			if(s[i] == '(')
				nodes.push(Node('(',i));
			else
			{
				Node top = nodes.top();
				if(top.ch == '(')
				{
					nodes.pop();
					ret = max(ret, i - nodes.top().index);
				}
				else
					nodes.push(Node(')',i));
			}
		}
		return ret;
        return 0;
    }
};

File: JAVA/LongestValidParentheses.java
import java.util.Stack;

class Solution {
	static class Node
	{
		public char ch;
		public int index;
		public Node(char c, int i)
		{
			ch = c;
			index = i;
		}
	}
    public int longestValidParentheses(String s) {
        // Start typing your Java solution below
        // DO NOT write main() function
        Stack<Node> stack = new Stack<Node>();
        stack.push(new Node(')', -1));
        int longest = 0;
        for(int i = 0 ; i < s.length(); i++)
        {
        	char ch = s.charAt(i);
        	if(ch == '(')
        	{
        		stack.push(new Node(ch, i));
        	}
        	else
        	{
        		Node n = stack.peek();
        		if(n.ch == '(')
        		{
        			stack.pop();
        			longest = Math.max(longest, i-stack.peek().index);
        		}
        		else
        			stack.push(new Node(ch, i));
        	}
        }
        return longest;
    }
}
public class LongestValidParentheses {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}
